proposals/interface-types/working-notes/scenarios/chooser.md

```diff
@@ -0,0 +1,680 @@
+          owned.release
+          i32.const 1                ;; initial reference count of 1
+          call-export "shared-builder-x" ;; make it a shared ptr pair
+          own (i32)
+            call-export "shared-release-x"
+          end
+        end
+        let (local $rp i32)(result i32)
+          local.get $lp
+          owned.access
+          local.get $rp
+          owned.access
+          call-export "nondeterministic_choice"  ;; call the chooser itself
+          own (i32)               ;; own the result
+            call-export "shared-release-x" 
+          end
+          owned.access 
+          call-export "shared-get-x"
+          call-export "access-utf8-x"
+          let (local $xbase i32) (local $xsize i32) (result i32)
+            local.get $xsize
+            call-export "malloc-i"     ;; create string entity of right size
+            let (local $ptr i32)
+              local.get $xbase
+              local.get $ptr
+              local.get $xsize
+              memory.copy "mem-x" "mem-i"
+              local.get $ptr           ;; our final return value
+            end
+          end
+        end
+      end
+    end
+  end
+)	
+```
+
+Our final step is unwrapping the `own`ed blocks and moving their contents to the
+correct place in the final adapter. This requires us to know where in the code
+the last reference to the owned value are.
+
+This is achieved in two phases: creating local variables that reference the
+stack values captured by the `own` instructions, and then moving the `own`ed
+block of instructions to the appropriate location.
+
+Note that the various `owned.access` instructions disappear at this point.
+
+In this case, the `$lp` and `$rp` values have no mention after the call to
+`"nondterministic_choice"`, and so we can move the `"shared-release"` calls to
+just after that call. Similarly, the release of the return value can be
+performed after returned string has been copied into the importing module:
+
+```wasm5
+(@interface implement (import "" "chooser_")
+  (param $l i32)
+  (param $r i32)
+  (result i32)
+  (local $o1 i32) ;; left string shared-ptr memory
+  (local $o2 i32) ;; right string shared-ptr memory
+  (local $o3 i32)
+
+  local.get $l
+  call-export "access-utf8-i" ;; return base & len
+  
+  let (local $lbase i32)(local $lsize i32)(result i32)
+    local.get $lbase
+    local.get $lsize
+    call-export "malloc-x"
+    let (local $ptr i32)(result i32)
+	  local.get $lbase
+	  local.get $ptr
+	  local.get $lsize
+      memory.copy "mem-i" "mem-x" ;; copy string across
+      local.get $ptr
+      i32.const 1                ;; initial reference count of 1
+      call-export "shared-builder-x" ;; make it a shared ptr pair
+      local.tee $o1
+    end
+    let (local $lp i32) (result i32)
+      local.get $r
+      call-export "access-utf8-i"
+      let (local $rbase i32)(local $rsize i32)(result i32)
+        local.get $rbase
+        local.get $rsize
+        call-export "malloc-x"
```

Comment:
it looks like we lost the matching `free-x` here...?

Reply:
the own.release cancels the own block.
if we were accounting for exceptions properly (which we are not yet) then the free-x call would should up in the handler.